src/AddressRegistry.Api.Legacy/AddressMatch/Matching/MunicipalityMatcher.cs

```diff
@@ -6,112 +6,155 @@ namespace AddressRegistry.Api.Legacy.AddressMatch.Matching
 
-        private IEnumerable<MunicipalityLatestItem> FilterByName(IEnumerable<MunicipalityLatestItem> municipalities, AddressMatchBuilder results)
+        protected override IReadOnlyList<TResult> BuildResultsInternal(AddressMatchBuilder? results)
+            => results
+                .Select(g => g.Municipality)
+                .Select(_mapper.Map)
+                .ToList();
+
+        private IEnumerable<MunicipalityLatestItem> FilterByName(
+            IEnumerable<MunicipalityLatestItem> municipalities,
+            AddressMatchBuilder? results)
         {
-            var municipalitiesByName = municipalities.Where(g => g.DefaultName.Value.EqIgnoreCase(results.Query.MunicipalityName)).ToList();
-            if (!string.IsNullOrEmpty(results.Query.MunicipalityName) && !municipalitiesByName.Any())
+            // TODO: Can't we skip this entire bit if there is no MunicipalityName query?
+
+            var municipalitiesByName = municipalities
+                .Where(g => g.DefaultName.Value.EqIgnoreCase(results?.Query.MunicipalityName))
+                .ToList();
+
+            if (!string.IsNullOrEmpty(results?.Query.MunicipalityName) && !municipalitiesByName.Any())
                 _warnings.AddWarning("6", "Onbekende 'Gemeentenaam'.");
 
             return municipalitiesByName;
         }
 
-        private IEnumerable<MunicipalityLatestItem> FilterByNisCode(IEnumerable<MunicipalityLatestItem> municipalities, AddressMatchBuilder results)
+        private IEnumerable<MunicipalityLatestItem> FilterByNisCode(
+            IEnumerable<MunicipalityLatestItem> municipalities,
+            AddressMatchBuilder? results)
         {
-            var municipalitiesByNisCode = municipalities.Where(g => g.NisCode.Equals(results.Query.NisCode)).ToList();
-            if (!string.IsNullOrEmpty(results.Query.NisCode) && !municipalitiesByNisCode.Any())
+            // TODO: Can't we skip this entire bit if there is no NisCode query?
+
+            var municipalitiesByNisCode = municipalities
+                .Where(g => !string.IsNullOrWhiteSpace(g.NisCode) && g.NisCode.Equals(results?.Query.NisCode))
+                .ToList();
+
+            if (!string.IsNullOrEmpty(results?.Query.NisCode) && !municipalitiesByNisCode.Any())
                 _warnings.AddWarning("5", "Onbekende 'Niscode'.");
 
             return municipalitiesByNisCode;
         }
 
-        private IEnumerable<MunicipalityLatestItem> FilterByPostalcode(IEnumerable<MunicipalityLatestItem> municipalities, AddressMatchBuilder results)
+        private IEnumerable<MunicipalityLatestItem> FilterByPostalCode(
+            IEnumerable<MunicipalityLatestItem> municipalities,
+            AddressMatchBuilder results)
         {
-            var postalInfo = _latestQueries.GetAllPostalInfo().FirstOrDefault(x => x.PostalCode == results.Query.PostalCode);
+            // TODO: Can't we skip this entire bit if there is no PostalCode query?
+
+            var postalInfo = _latestQueries
```

Comment:
Can't we skip this entire bit if there is no `PostalCode` query?

Reply:
yep